From 4f35176acaab12f9e2bbfba4be60440abdc65cb3 Mon Sep 17 00:00:00 2001
From: Vishal Sagar <vishal.sagar@xilinx.com>
Date: Tue, 22 Sep 2020 12:24:20 -0700
Subject: [PATCH 1644/1851] v4l: xilinx: cfa: Fix warnings for NULL_RETURNS

commit 2910605c14a95f144ae09a40c8a9c34a5aa5e3ae from
https://github.com/Xilinx/linux-xlnx.git

This patch fix the warnings for the NULL_RETURNS type.

The actual warnings are:

For __xcfa_get_pad_format():
Event return_null: Explicitly returning null.

For xcfa_get_format():
Event returned_null: 	"__xcfa_get_pad_format" returns "NULL".
Event dereference: 	Dereferencing "__xcfa_get_pad_format(xcfa, cfg,
fmt->pad, fmt->which)", which is known to be "NULL"

For xcfa_set_format():
Event returned_null: 	"__xcfa_get_pad_format" returns "NULL".
Event var_assigned: 	Assigning: "format" = "NULL" return value from
"__xcfa_get_pad_format".
Event dereference: 	Dereferencing "format", which is known to be
"NULL".

Signed-off-by: Vishal Sagar <vishal.sagar@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/media/platform/xilinx/xilinx-cfa.c | 23 ++++++++++++++++++----
 1 file changed, 19 insertions(+), 4 deletions(-)

diff --git a/drivers/media/platform/xilinx/xilinx-cfa.c b/drivers/media/platform/xilinx/xilinx-cfa.c
index 832fb7306563..e71a4795ee36 100644
--- a/drivers/media/platform/xilinx/xilinx-cfa.c
+++ b/drivers/media/platform/xilinx/xilinx-cfa.c
@@ -108,14 +108,22 @@ __xcfa_get_pad_format(struct xcfa_device *xcfa,
 		      struct v4l2_subdev_pad_config *cfg,
 		      unsigned int pad, u32 which)
 {
+	struct v4l2_mbus_framefmt *format;
+
 	switch (which) {
 	case V4L2_SUBDEV_FORMAT_TRY:
-		return v4l2_subdev_get_try_format(&xcfa->xvip.subdev, cfg, pad);
+		format = v4l2_subdev_get_try_format(&xcfa->xvip.subdev, cfg,
+						    pad);
+		break;
 	case V4L2_SUBDEV_FORMAT_ACTIVE:
-		return &xcfa->formats[pad];
+		format = &xcfa->formats[pad];
+		break;
 	default:
-		return NULL;
+		format = NULL;
+		break;
 	}
+
+	return format;
 }
 
 static int xcfa_get_format(struct v4l2_subdev *subdev,
@@ -123,8 +131,13 @@ static int xcfa_get_format(struct v4l2_subdev *subdev,
 			   struct v4l2_subdev_format *fmt)
 {
 	struct xcfa_device *xcfa = to_cfa(subdev);
+	struct v4l2_mbus_framefmt *format;
 
-	fmt->format = *__xcfa_get_pad_format(xcfa, cfg, fmt->pad, fmt->which);
+	format = __xcfa_get_pad_format(xcfa, cfg, fmt->pad, fmt->which);
+	if (!format)
+		return -EINVAL;
+
+	fmt->format = *format;
 
 	return 0;
 }
@@ -138,6 +151,8 @@ static int xcfa_set_format(struct v4l2_subdev *subdev,
 	int bayer_phase;
 
 	format = __xcfa_get_pad_format(xcfa, cfg, fmt->pad, fmt->which);
+	if (!format)
+		return -EINVAL;
 
 	if (fmt->pad == XVIP_PAD_SOURCE) {
 		fmt->format = *format;
-- 
2.31.1

